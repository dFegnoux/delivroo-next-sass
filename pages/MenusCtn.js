import React, { Component, Fragment } from "react";
import axios from "axios";
import RestaurantBlock from "../components/blocks/RestaurantBlock";
import PacmanLoader from "react-spinners/PacmanLoader";
import Menu from "../components/menu/Menu";
import { formatMenus } from "../helpers/menuHelpers";
import "../components/menu/menu.scss";

class CartValidationCtn extends Component {
  state = {
    cart: {},
    showPayment: false
  };

  // Server side et avant le premier render
  static async getInitialProps() {
    const response = await axios.get("https://deliveroo-api.now.sh/menu");

    if (response.data instanceof Object) {
      return {
        menu: formatMenus(response.data.menu),
        restaurant: response.data.restaurant || {}
      };
    }

    return {
      menu: [],
      restaurant: {}
    };
  }

  /**
   * Save the cart to local storage,
   * cart item is provided by the component's state
   */
  saveCartToStorage = () => {
    localStorage.setItem("cart", JSON.stringify(this.state.cart));
  };

  /**
   * Add, remove menus from cart
   * @param {Object} menu : A menu object from delivroo API
   * @param {String} action : 'add' or 'remove'
   */
  updateCart = (menu, action) => {
    // Get previous item if it already exists
    const previousItem = this.state.cart.hasOwnProperty(menu.id)
      ? this.state.cart[menu.id]
      : null;

    let newQuantity = previousItem ? previousItem.quantity : 0;
    if (action === "remove") {
      newQuantity--;
    } else {
      newQuantity++;
    }

    if (newQuantity) {
      this.setState(
        () => ({
          cart: {
            ...this.state.cart,
            [menu.id]: {
              ...menu,
              name: menu.title,
              quantity: newQuantity
            }
          }
        }),
        this.saveCartToStorage
      );
    } else {
      const newCart = {
        ...this.state.cart
      };

      delete newCart[menu.id];

      this.setState(() => ({ cart: newCart }), this.saveCartToStorage);
    }
  };

  componentDidMount() {
    const savedCart = localStorage.getItem("cart");
    if (savedCart) {
      this.setState(() => ({ cart: JSON.parse(savedCart) }));
    }
  }

  render() {
    const { cart, showPayment } = this.state;
    const { menu, restaurant } = this.props;

    return (
      <Fragment>
        {Object.keys(restaurant).length > 0 && (
          <RestaurantBlock
            name={restaurant.name}
            description={restaurant.description}
            picture={restaurant.picture}
          />
        )}
        <div className="layoutPageContent layoutLimitedWidth">
          {!menu.length ? (
            <PacmanLoader />
          ) : (
            <Menu
              categories={menu}
              updateCart={this.updateCart}
              cart={cart}
              showPayment={showPayment}
            />
          )}
        </div>
      </Fragment>
    );
  }
}

export default CartValidationCtn;
