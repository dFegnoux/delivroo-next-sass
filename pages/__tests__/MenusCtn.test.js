import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import MenusCtn from "../MenusCtn";
import mockAxios from "jest-mock-axios";

const mandatoryProps = {
  menu: [],
  restaurant: {}
};

describe("MenusCtn snapshots", () => {
  const filledProps = {
    restaurant: {
      name: "Mocked restaurant",
      description: "Mocked restaurant description",
      picture: "Mocked restaurant picture"
    },
    menu: [["category1", [{ id: "mockedId1", description: "" }]]]
  };
  it("should render correctly", () => {
    const tree = renderer.create(<MenusCtn {...mandatoryProps} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should render correctly with a restaurant and menus", () => {
    const tree = renderer.create(<MenusCtn {...filledProps} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe("MenusCtn lifecycle", () => {
  afterEach(() => {
    localStorage.removeItem("cart");
  });
  it("Should retrieve cart from localStorage", () => {
    const cartToRetrieve = {
      storageCartItemId: { title: "storage item", quantity: 1 }
    };
    localStorage.setItem("cart", JSON.stringify(cartToRetrieve));
    const container = shallow(<MenusCtn {...mandatoryProps} />);
    expect(container.state("cart")).toEqual(cartToRetrieve);
  });
});

describe("MenusCtn getInitialProps", () => {
  afterEach(() => {
    // cleaning up the mess left behind the previous test
    mockAxios.reset();
  });

  it("Should set values retrieved by axios", async () => {
    MenusCtn.getInitialProps();
    mockAxios.mockResponse({ data: { restaurant: {}, menu: {} } });

    expect(mockAxios.get).toHaveBeenCalledWith(
      "https://deliveroo-api.now.sh/menu"
    );
  });

  it("Should handle when call returns no restaurant", async () => {
    MenusCtn.getInitialProps();
    mockAxios.mockResponse({ data: { menu: {} } });

    expect(mockAxios.get).toHaveBeenCalledWith(
      "https://deliveroo-api.now.sh/menu"
    );
  });

  it("Should set default values when axios answer isn't satisfying", async () => {
    MenusCtn.getInitialProps();
    mockAxios.mockResponse({ data: "" });

    expect(mockAxios.get).toHaveBeenCalledWith(
      "https://deliveroo-api.now.sh/menu"
    );
  });
});

describe("MenusCtn functions", () => {
  const container = shallow(<MenusCtn {...mandatoryProps} />);
  const CtnInstance = container.instance();
  const newItem = { title: "my mocked item", id: "mockedId" };
  it("should be able to save in localStorage", () => {
    CtnInstance.saveCartToStorage();
    expect(localStorage.setItem).toHaveBeenCalledWith("cart", "{}");
  });

  it("should add an item to cart", () => {
    CtnInstance.updateCart(newItem, "add");
    const cartState = container.state("cart");
    expect(Object.keys(cartState).length).toBe(1);
    expect(Object.keys(cartState)[0]).toBe(newItem.id);
    expect(cartState[newItem.id].quantity).toBe(1);
  });

  it("should update quantity of an existing item to cart", () => {
    CtnInstance.updateCart(newItem, "add");
    expect(container.state("cart")[newItem.id].quantity).toBe(2);

    CtnInstance.updateCart(newItem, "remove");
    expect(container.state("cart")[newItem.id].quantity).toBe(1);
  });

  it("should remove an item to cart when its quantity drop to zero", () => {
    CtnInstance.updateCart(newItem, "remove");
    const cartState = container.state("cart");
    expect(Object.keys(cartState).length).toBe(0);
  });
});
