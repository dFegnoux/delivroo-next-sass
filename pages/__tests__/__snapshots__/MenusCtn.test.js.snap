// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenusCtn snapshots should render correctly 1`] = `
<div
  className="layoutPageContent layoutLimitedWidth"
>
  <div
    className="css-ph4qhg"
  >
    <div
      className="css-10zneet"
    />
    <div
      className="css-18rqjra"
    />
    <div
      className="css-vyfd34"
    />
    <div
      className="css-3syn48"
    />
    <div
      className="css-18rniua"
    />
    <div
      className="css-ueou2"
    />
  </div>
</div>
`;

exports[`MenusCtn snapshots should render correctly with a restaurant and menus 1`] = `
Array [
  <div
    className="restaurantBlockWrapper centerContent"
  >
    <div
      className="restaurantBlockContent resetTextAlign"
    >
      <div
        className="texts"
      >
        <div
          className="newsTitle"
        >
          Mocked restaurant
        </div>
        <div
          className="newsDescription"
        >
          Mocked restaurant description
        </div>
      </div>
      <div
        className="newsPictureWrapper"
      >
        <img
          alt="Mocked restaurant"
          src="Mocked restaurant picture"
        />
      </div>
    </div>
  </div>,
  <div
    className="layoutPageContent layoutLimitedWidth"
  >
    <div
      className="menuLayout"
    >
      <div
        className="menuContentWrapper"
      >
        <div
          className="menuCategory"
        >
          <div
            className="title"
          >
            category1
          </div>
          <div
            className="itemsList"
          >
            <div
              className="menuCard"
              onClick={[Function]}
            >
              <div
                className="content roundedItem"
              >
                <div
                  className="texts"
                >
                  <div
                    className="cardTitle"
                  />
                  <div
                    className="description"
                  >
                    
                  </div>
                  <div
                    className="bottom"
                  >
                    <span
                      className="price"
                    >
                       â‚¬
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="cartWrapper"
      >
        <div
          className="cart roundedItem"
        >
          <a
            className="btn"
            disabled="disabled"
            href="/cart/validation"
            onClick={[Function]}
          >
            Valider mon panier
          </a>
          <div>
            Votre panier est vide
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
